sdk/eventhub/eventhubs-checkpointstore-table/src/tableCheckpointStore.ts

```diff
@@ -42,13 +109,69 @@ export class TableCheckpointStore implements CheckpointStore {
    *  - `tracingOptions`: Options for configuring tracing.
    * @returns A list partitions this instance successfully claimed ownership.
    */
-  async claimOwnership(): /*
-    partitionOwnership: PartitionOwnership[],
-    options: OperationOptions = {}
-    */
-  Promise<PartitionOwnership[]> {
-    console.log("nothing to claim");
-    throw new Error("not implemented");
+  async claimOwnership(partitionOwnership: PartitionOwnership[]): Promise<PartitionOwnership[]> {
+    const partitionOwnershipArray: PartitionOwnership[] = [];
+
+    for (const ownership of partitionOwnership) {
+      const partition_Key = `${ownership.fullyQualifiedNamespace} ${ownership.eventHubName} ${ownership.consumerGroup} Ownership`;
+      const ownershipEntity: PartitionOwnershipEntity = {
+        partitionKey: partition_Key,
+        rowKey: ownership.partitionId,
+        ownerid: ownership.ownerId
+      };
+
+      // When we have an etag, we know the entity existed.
+      // If we encounter an error we should fail.
+      try {
+        if (ownership.etag) {
+          const updatedMetadata = await this._tableClient.updateEntity(ownershipEntity, "Replace", {
+            etag: ownership.etag
+          });
+          ownership.etag = updatedMetadata.etag;
+          ownership.lastModifiedTimeInMs = await this._toMillisecs(
+            (updatedMetadata as any).Timestamp
+          );
+          partitionOwnershipArray.push(ownership);
+          logger.info(
+            `[${ownership.ownerId}] Claimed ownership successfully for partition: ${ownership.partitionId}`,
+            `LastModifiedTime: ${ownership.lastModifiedTimeInMs}, ETag: ${ownership.etag}`
+          );
+        } else {
+          try {
+            const newOwnershipMetadata = await this._tableClient.createEntity(ownershipEntity, {
+              requestOptions: {
+                customHeaders: {
+                  Prefer: "return-content"
+                }
+              }
+            });
+            ownership.etag = newOwnershipMetadata.etag;
+            ownership.lastModifiedTimeInMs = await this._toMillisecs(
+              (newOwnershipMetadata as any).Timestamp
+            );
```

Comment:
Today, we're going to learn about Type Guards!

Right now, you have to cast `newOwnershipMetadata` to `any` to access `Timestamp`, otherwise the compiler complains. However, we lose type safety by doing this, oh no!

Well, TypeScript has a handy feature called a type guard that helps us get some type safety back! A type guard is simply a function that returns `true` or `false`. If it returns `true`, then the function asserts something about the input.

Here's the type guard you'll eventually be able to use:
```ts
function hasTimestamp<T extends TableInsertEntityHeaders>(value: T): value is T & { Timestamp: string } {
  return typeof (value as any).Timestamp === 'string';
}
```
Let's break it down: `<T extends TableInsertEntityHeaders>(value: T)`

This says that the function accepts any input for `value` that structurally matches `TableInsertEntityHeaders` - the output of `createEntity`.

Now the return type: `value is T & { Timestamp: string }`

This return type means that if the function returns `true`, then the input, `value`, is a combination of `T` (`TableInsertEntityHeaders`) and `{ Timestamp: string }`.

Finally, we do a runtime type check of the `value` to test if `Timestamp` exists.

Now, you can write code like this:
```ts
if (!hasTimestamp(newOwnershipMetadata)) {
  throw new Error(`Unable to get the timestamp...etc etc`);
}
updatedOwnership.lastModifiedTimeInMs = new Date(newOwnershipMetadata.Timestamp).getTime();
```
Notice that I don't have to cast `newOwnershipMetadata` to `any` anymore!

Reply:
yes this looks good